xarray/core/variable.py

```diff
@@ -197,6 +190,10 @@ def as_compatible_data(data, fastpath=False):
         else:
             data = np.asarray(data)
 
+    # validate whether the data is valid data types
+    if not isinstance(data, SUPPORT_ARRAY_TYPES):
+        data = np.asarray(data)
```

Comment:
I think we should check `data` is an array-like and raise an Error if not.
(`np.asarray` converts any object to a `np.ndarray`.)

Maybe we need to follow [this](https://stackoverflow.com/questions/40378427/numpy-formal-definition-of-array-like-objects) definition of `array-like`?
Do we already have any utility function for this?